proxy/src/main/java/com/wavefront/agent/TenantInfo.java

```diff
@@ -0,0 +1,164 @@
+      @Nonnull final String clientSecret,
+      @Nonnull final String orgId,
+      @Nonnull final String server) {
+    this.server = server;
+    loadAccessTokenByClientCredentials(clientId, clientSecret, orgId);
+  }
+
+  @Nonnull
+  public String getServer() {
+    return server;
+  }
+
+  @Nonnull
+  public String getToken() {
+    return bearerToken;
+  }
+
+  /**
+   * When the CSP access token has expired, obtain it using the CSP api token.
+   *
+   * @param apiToken the CSP api token.
+   * @throws RuntimeException when CSP is down or wrong CSP api token.
+   */
+  private void loadAccessTokenByAPIToken(@Nonnull final String apiToken) {
+
+    Form requestBody = new Form();
+    requestBody.param("grant_type", "api_token");
+    requestBody.param("api_token", apiToken);
+
+    Response response =
+        new ResteasyClientBuilder()
+            .build()
+            .target(format("%s/csp/gateway/am/api/auth/api-tokens/authorize", cspBaseUrl))
+            .request()
+            .post(Entity.form(requestBody));
+
+    if (response.getStatusInfo().getFamily() != SUCCESSFUL) {
+      throw new RuntimeException(
+          GET_CSP_ACCESS_TOKEN_ERROR_MESSAGE + " Status: " + response.getStatus());
+    }
+
+    final TokenExchangeResponseDTO tokenExchangeResponseDTO =
+        response.readEntity(TokenExchangeResponseDTO.class);
+
+    if (tokenExchangeResponseDTO == null) {
+      throw new RuntimeException(GET_CSP_ACCESS_TOKEN_ERROR_MESSAGE + " Response object is empty.");
+    }
+
+    this.bearerToken = tokenExchangeResponseDTO.getAccessToken();
+    executor.schedule(
+        () -> loadAccessTokenByAPIToken(apiToken),
+        tokenExchangeResponseDTO.getExpiresIn(),
+        TimeUnit.SECONDS);
+  }
+
+  /**
+   * When the CSP access token has expired, obtain it using server by server app url and client
+   * credentials.
+   *
+   * @param clientId the CSP OAuth app client id.
+   * @param clientSecret the CSP OAuth app client secret.
+   * @param orgId the CSP organisation id.
+   * @throws RuntimeException when CSP is down or wrong csp client credentials.
+   */
+  private void loadAccessTokenByClientCredentials(
+      @Nonnull final String clientId,
+      @Nonnull final String clientSecret,
+      @Nonnull final String orgId) {
+
+    Form requestBody = new Form();
+    requestBody.param("grant_type", "client_credentials");
+    requestBody.param("orgId", orgId);
+
+    Response response =
+        new ResteasyClientBuilder()
+            .build()
+            .target(format("%s/csp/gateway/am/api/auth/authorize", cspBaseUrl))
+            .request()
+            .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_FORM_URLENCODED)
+            .header(
+                HttpHeaders.AUTHORIZATION,
+                buildClientAuthorizationHeaderValue(clientId, clientSecret))
+            .post(Entity.form(requestBody));
+
+    if (response.getStatusInfo().getFamily() != SUCCESSFUL) {
+      throw new RuntimeException(
```

Comment:
this will prevent the `loadAccessTokenByClientCredentials` to be scheduled.

Reply:
Yes, it is the happy way version. Another ticket will be created to have better error handling. Additionally, I'll add unit tests during that ticket before putting it up for review.

Reply:
why another ticket ? is this PR just for testing?